<a href="https://colab.research.google.com/github/mucahituzak/KampIntro/blob/master/VeriMadenciligiFinalEvFiyatlariTahmini.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Veri Kümesini Yükleyelim**

---

İlk adım indirdiğimiz csv dosyasından veri kümesini yüklemek.Dosyayı yerel sürücüden yüklemek için Google Colap paketinden files.upload()fonksiyonunu kullanacağız:

In [2]:
from google.colab import files


uploaded = files.upload()


Saving AmesHousing.csv to AmesHousing (1).csv


**Veri Kümesini Okuyalım**

---

Yüklediğimiz csv dosyasından veri kümesini okuyacağız. Pandas Python paketinden read_csv() fonksiyonunu kullanacağız:


In [3]:
import pandas as pd
import numpy as np

dataset = pd.read_csv("AmesHousing.csv")

**Veri Setini Anlayalım**

---
Bir fikir edinmek için veri kümesinin ilk birkaç satırını görüntüleyelim:


In [4]:
# Kayan sayı formatını yapılandırma
pd.options.display.float_format = '{:20.2f}'.format
dataset.head(n=5)


Order        PID  MS SubClass MS Zoning         Lot Frontage  Lot Area  \
0      1  526301100           20        RL               141.00     31770   
1      2  526350040           20        RH                80.00     11622   
2      3  526351010           20        RL                81.00     14267   
3      4  526353030           20        RL                93.00     11160   
4      5  527105010           60        RL                74.00     13830   

  Street Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence  \
0   Pave   NaN       IR1          Lvl  ...         0     NaN    NaN   
1   Pave   NaN       Reg          Lvl  ...         0     NaN  MnPrv   
2   Pave   NaN       IR1          Lvl  ...         0     NaN    NaN   
3   Pave   NaN       Reg          Lvl  ...         0     NaN    NaN   
4   Pave   NaN       IR1          Lvl  ...         0     NaN  MnPrv   

  Misc Feature Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
0          NaN        0       5    2010       WD           Normal     215000  
1          NaN        0       6    2010       WD           Normal     105000  
2         Gar2    12500       6    2010       WD           Normal     172000  
3          NaN        0       4    2010       WD           Normal     244000  
4          NaN        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

Şimdi, veri kümemizdeki sayısal sütunlar hakkında istatistiksel bilgi alalım. Veri kümesindeki her bir sayısal sütun için ortalama, standart sapma, minimum, maksimum ve 50. yüzdelik dilimi (medyan) bilmek istiyoruz:

In [7]:
dataset.describe(include=[np.number], percentiles=[.5]) \
    .transpose().drop("count", axis=1)


mean                  std  \
Order                        1465.50               845.96   
PID                     714464496.99         188730844.65   
MS SubClass                    57.39                42.64   
Lot Frontage                   69.22                23.37   
Lot Area                    10147.92              7880.02   
Overall Qual                    6.09                 1.41   
Overall Cond                    5.56                 1.11   
Year Built                   1971.36                30.25   
Year Remod/Add               1984.27                20.86   
Mas Vnr Area                  101.90               179.11   
BsmtFin SF 1                  442.63               455.59   
BsmtFin SF 2                   49.72               169.17   
Bsmt Unf SF                   559.26               439.49   
Total Bsmt SF                1051.61               440.62   
1st Flr SF                   1159.56               391.89   
2nd Flr SF                    335.46               428.40   
Low Qual Fin SF                 4.68                46.31   
Gr Liv Area                  1499.69               505.51   
Bsmt Full Bath                  0.43                 0.52   
Bsmt Half Bath                  0.06                 0.25   
Full Bath                       1.57                 0.55   
Half Bath                       0.38                 0.50   
Bedroom AbvGr                   2.85                 0.83   
Kitchen AbvGr                   1.04                 0.21   
TotRms AbvGrd                   6.44                 1.57   
Fireplaces                      0.60                 0.65   
Garage Yr Blt                1978.13                25.53   
Garage Cars                     1.77                 0.76   
Garage Area                   472.82               215.05   
Wood Deck SF                   93.75               126.36   
Open Porch SF                  47.53                67.48   
Enclosed Porch                 23.01                64.14   
3Ssn Porch                      2.59                25.14   
Screen Porch                   16.00                56.09   
Pool Area                       2.24                35.60   
Misc Val                       50.64               566.34   
Mo Sold                         6.22                 2.71   
Yr Sold                      2007.79                 1.32   
SalePrice                  180796.06             79886.69   

                                 min                  50%                  max  
Order                           1.00              1465.50              2930.00  
PID                     526301100.00         535453620.00        1007100110.00  
MS SubClass                    20.00                50.00               190.00  
Lot Frontage                   21.00                68.00               313.00  
Lot Area                     1300.00              9436.50            215245.00  
Overall Qual                    1.00                 6.00                10.00  
Overall Cond                    1.00                 5.00                 9.00  
Year Built                   1872.00              1973.00              2010.00  
Year Remod/Add               1950.00              1993.00              2010.00  
Mas Vnr Area                    0.00                 0.00              1600.00  
BsmtFin SF 1                    0.00               370.00              5644.00  
BsmtFin SF 2                    0.00                 0.00              1526.00  
Bsmt Unf SF                     0.00               466.00              2336.00  
Total Bsmt SF                   0.00               990.00              6110.00  
1st Flr SF                    334.00              1084.00              5095.00  
2nd Flr SF                      0.00                 0.00              2065.00  
Low Qual Fin SF                 0.00                 0.00              1064.00  
Gr Liv Area                   334.00              1442.00              5642.00  
Bsmt Full Bath                  0.00                

Yukarıdaki tablodan, örneğin, veri setimizdeki evlerin ortalama arsa alanının 10.147,92 ft2 olduğunu ve standart sapmanın 7.880,02 ft2 olduğunu görebiliriz. Ayrıca minimum arsa alanının 1.300 ft2 ve maksimum arsa alanının 215.245 ft2 olduğunu ve medyan değerinin 9.436,5 ft2 olduğunu görebiliriz. Benzer şekilde, tablodan veri kümemizdeki değişkenler hakkında pek çok bilgi edinebiliriz.

---

Ardından, veri setimizdeki sayısal olmayan sütunlarla ilgili istatistiksel bilgileri görmeye geçiyoruz:




In [13]:
dataset.describe(include=[object]).transpose() \


count unique      top  freq
MS Zoning       2930      7       RL  2273
Street          2930      2     Pave  2918
Alley            198      2     Grvl   120
Lot Shape       2930      4      Reg  1859
Land Contour    2930      4      Lvl  2633
Utilities       2930      3   AllPub  2927
Lot Config      2930      5   Inside  2140
Land Slope      2930      3      Gtl  2789
Neighborhood    2930     28    NAmes   443
Condition 1     2930      9     Norm  2522
Condition 2     2930      8     Norm  2900
Bldg Type       2930      5     1Fam  2425
House Style     2930      8   1Story  1481
Roof Style      2930      6    Gable  2321
Roof Matl       2930      8  CompShg  2887
Exterior 1st    2930     16  VinylSd  1026
Exterior 2nd    2930     17  VinylSd  1015
Mas Vnr Type    2907      5     None  1752
Exter Qual      2930      4       TA  1799
Exter Cond      2930      5       TA  2549
Foundation      2930      6    PConc  1310
Bsmt Qual       2850      5       TA  1283
Bsmt Cond       2850      5       TA  2616
Bsmt Exposure   2847      4       No  1906
BsmtFin Type 1  2850      6      GLQ   859
BsmtFin Type 2  2849      6      Unf  2499
Heating         2930      6     GasA  2885
Heating QC      2930      5       Ex  1495
Central Air     2930      2        Y  2734
Electrical      2929      5    SBrkr  2682
Kitchen Qual    2930      5       TA  1494
Functional      2930      8      Typ  2728
Fireplace Qu    1508      5       Gd   744
Garage Type     2773      6   Attchd  1731
Garage Finish   2771      3      Unf  1231
Garage Qual     2771      5       TA  2615
Garage Cond     2771      5       TA  2665
Paved Drive     2930      3        Y  2652
Pool QC           13      4       Ex     4
Fence            572      4    MnPrv   330
Misc Feature     106      5     Shed    95
Sale Type       2930     10      WD   2536
Sale Condition  2930      6   Normal  2413

Elimizdeki tabloda, count her sütundaki boş olmayan değerlerin sayısını, unique benzersiz değerlerin sayısını, top en sık kullanılan öğeyi ve freq en sık kullanılan öğenin sıklığını temsil eder.

**Veri Temizleme**

---
**Eksik Değerlerle Başa Çıkma**


Bazı makine öğrenimi modelleri eksik değerlere sahip verileri kabul etmediği için eksik değerler sorunuyla ilgilenmeliyiz. İlk olarak, veri kümemizdeki eksik değerlerin sayısını görelim. Gerçekte eksik değerler içeren her sütun için eksik değerlerin sayısını ve yüzdesini görmek istiyoruz.

In [14]:
# Her sütundaki eksik değerlerin sayısını alma
num_missing = dataset.isna().sum()
# Eksik 0 değer içeren sütunların hariç tutulması
num_missing = num_missing[num_missing > 0]
# Kayıp değerlerin yüzdelerini alma
percent_missing = num_missing * 100 / dataset.shape[0]
# Kayıp değerlerin sayısı ve yüzdesinin birleştirilmesi
# tek bir veri çerçevesi haline getirme ve sıralama
pd.concat([num_missing, percent_missing], axis=1,
          keys=['Missing Values', 'Percentage']).\
          sort_values(by="Missing Values", ascending=False)

Missing Values           Percentage
Pool QC                   2917                99.56
Misc Feature              2824                96.38
Alley                     2732                93.24
Fence                     2358                80.48
Fireplace Qu              1422                48.53
Lot Frontage               490                16.72
Garage Cond                159                 5.43
Garage Qual                159                 5.43
Garage Finish              159                 5.43
Garage Yr Blt              159                 5.43
Garage Type                157                 5.36
Bsmt Exposure               83                 2.83
BsmtFin Type 2              81                 2.76
BsmtFin Type 1              80                 2.73
Bsmt Qual                   80                 2.73
Bsmt Cond                   80                 2.73
Mas Vnr Area                23                 0.78
Mas Vnr Type                23                 0.78
Bsmt Half Bath               2                 0.07
Bsmt Full Bath               2                 0.07
Total Bsmt SF                1                 0.03
Bsmt Unf SF                  1                 0.03
Garage Cars                  1                 0.03
Garage Area                  1                 0.03
BsmtFin SF 2                 1                 0.03
BsmtFin SF 1                 1                 0.03
Electrical                   1                 0.03

Şimdi bu kayıp değerlerle uğraşmaya başlıyoruz.

**Havuz Kalite Kontrol**

---

Havuz  sütunundaki eksik değerlerin yüzdesi %99,56'dır ve bu çok yüksek bir orandır. Bu sütundaki eksik değerin ilgili evin havuzu olmadığını gösterdiğini düşünüyoruz. Bunu doğrulamak için Havuz Alanı sütunundaki değerlere bir göz atalım:

In [15]:
dataset["Pool Area"].value_counts()

0      2917
144       1
480       1
576       1
555       1
368       1
444       1
228       1
561       1
519       1
648       1
800       1
512       1
738       1
Name: Pool Area, dtype: int64

Havuz Alanı sütununda 0 değerine sahip 2917 girdi olduğunu görebiliriz. Bu, havuzu olmayan her evin Havuz QC sütununda eksik bir değere ve Havuz Alanı sütununda 0 değerine sahip olduğu hipotezimizi doğrulamaktadır. O halde Havuz QC sütunundaki eksik değerleri "Havuz Yok" ile dolduralım:

In [16]:
dataset["Pool QC"].fillna("No Pool", inplace=True)